src/main/java/org/graylog/collector/file/PathSet.java

```diff
@@ -0,0 +1,195 @@
+                Files.walkFileTree(basePath, visitor);
+            }
+        });
+    }
+
+    public PathSet(final String pattern, FileTreeWalker fileTreeWalker) {
+        this.pattern = pattern;
+        this.fileTreeWalker = fileTreeWalker;
+        final String patternString = Utils.isWindows() ? pattern.replace("\\", "\\\\") : pattern;
+        this.matcher = FileSystems.getDefault().getPathMatcher("glob:" + patternString);
+        this.rootPath = getRootPathFromPattern(pattern);
+    }
+
+    public Path getRootPath() {
+        return rootPath;
+    }
+
+    public String getPattern() {
+        return pattern;
+    }
+
+    public boolean isInSet(Path path) {
+        return matcher.matches(path);
+    }
+
+    /**
+     * Returns a {@link Set<Path>} of all existing paths that match the pattern.
+     *
+     * The file tree is walked on every invocation to pick up newly created paths. This can be expensive!
+     *
+     * @return all existing paths that match the pattern
+     * @throws IOException
+     */
+    public Set<Path> getPaths() throws IOException {
+        final ImmutableSet.Builder<Path> matchedPaths = ImmutableSet.builder();
+
+        if (hasGlobMetaChars(pattern)) {
+            fileTreeWalker.walk(rootPath, new GlobbingFileVisitor(matcher, matchedPaths));
+        } else {
+            // TODO needs to be an absolute path because otherwise the FileObserver does weird things. Investigate what's wrong with it.
+            matchedPaths.add(Paths.get(pattern).toAbsolutePath());
+        }
+
+        return matchedPaths.build();
+    }
+
+    private Path getRootPathFromPattern(String pattern) {
+        if (!hasGlobMetaChars(pattern)) {
+            return Paths.get(pattern).toAbsolutePath().getParent();
+        }
+
+        final List<String> elements = Lists.newArrayList();
+
+        // Splitting manually on File.separator here to make it work on Windows.
+        for (String part : Splitter.on(File.separator).split(pattern)) {
+            if (hasGlobMetaChars(part)) {
+                break;
+            }
+
+            elements.add(part);
+        }
+
+        if (Utils.isWindows()) {
+            return Paths.get("", elements.toArray(new String[elements.size()]));
+        } else {
+            return Paths.get("/", elements.toArray(new String[elements.size()]));
+        }
+
+    }
+
+    private boolean hasGlobMetaChars(String path) {
+        if (path == null) {
+            return false;
+        }
+        for (String globMetaChar : GLOB_META_CHARS) {
+            if (path.contains(globMetaChar)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        PathSet pathSet = (PathSet) o;
+
+        return pattern.equals(pathSet.pattern);
```

Comment:
This will fail with an NPE if `pattern` is `null` (which isn't checked in the constructor).